e3nn/io/_cartesian_tensor.py

```diff
@@ -1,10 +1,46 @@
 from typing import Optional
+from threading import local
+from collections import defaultdict
+import copy
 
 import torch
 
 from e3nn import o3
 
 
+# CartesianTensor shouldn't carry a torch.nn.Module child
+# but we don't want to rebuild the RTP every call
+# cache it here in thread local storage (to avoid
+# possibility of weird threading bugs)
+_RTP_CACHE = local()
+_RTP_CACHE.cache = dict()  # Dict[(formula, device, dtype), ReducedTensorProducts]
```

Comment:
this is a good idea— I completely forgot `weakref` somehow while writing this— but it would be `WeakKeyDictionary`, no? Keyed by the `CartesianTensor` object? Each `CartesianTensor` is a new object, so this would not share across formulas...:
```
In [1]: from e3nn.io import CartesianTensor

In [2]: ct = CartesianTensor("ij=ji")

In [3]: id(ct)
Out[3]: 140139331168592

In [4]: ct2 = CartesianTensor("ij=ji")

In [5]: id(ct2)
Out[5]: 140139321106240

In [6]: id(ct) != id(ct2)
Out[6]: True
```
but I think you're right that there has to be a better way to do this with `weakref`

Reply:
Basically we want to have no strong references left when there are no live objects with a given _formula_.

`CartesianTensor` objects are tuples... we could make them singletons and then do the RTP cache as a `WeakKeyDict`? @mariogeiger ?

Reply:
Yes, I think this could be pushed all the way up to `RTP`.  Basically any immutable object that could be reused should be a singleton (probably through a `WeakValueDict` I would have thought), similar to how integers in python are [singletons](https://www.codementor.io/@arpitbhayani/python-caches-integers-16jih595jk) and everyone using, say, the number `2` is actually accessing the same instance in memory.

Sorry, maybe I'm missing something but I thought `WeakValueDict` because your current signature is:
```Dict[(formula, device, dtype), ReducedTensorProducts]```
In which case the keys of this would be `(formula, device, dtype)` which would define a one-to-one mapping to the corresponding `ReducedTensorProduct` which is then weakly referenced.  When all shared references to it disappear, then the dict would automatically pop that RTP.

Reply:
@Linux-cpp-lisp  Not sure how making `CartesianTensor` singletons would help here.
@muhrin if `ReducedTensorProduct` is weakly referenced in the cache it has to be strongly referenced elsewhere otherwise it will be deleted. And we don't want `CartesianTensor` to contain references to any `ReducedTensorProduct`

Reply:
So my thinking is that `CartesianTensor` has the strong ref to its `ReducedTensorProduct` instance and it's up to the user to keep the strong ref to the `CartesianTensor`.  I think this scheme would replicate the behaviour that Alby has here (currently this behaviour is achieved through the `_RTP_CACHE` in `_make_rtp`).

Reply:
We don't want `CartesianTensor` to have any ref to `ReducedTensorProduct` because two reasons:

1. `CartesianTensor` is an `Irreps` so it should not contain data
2. `CartesianTensor` is not an `nn.Module` so it cannot support multiple `dtype` and `device` (it has no `.to()` method)

Reply:
The strategy is therefore to create/recover a `ReducedTensorProduct` with the valid `detype/device` based on the `dtype/device` of the input data of the method used.

Reply:
Ah, I see, there can be multiple `ReducedTensorProduct`s per `CartesianTensor`, now I understand Alby's response from earlier...

Ok, so indeed to make this work as I had in mind `CartesianTensor` would have to be a singleton per formula and then there could be a `WeakKeyDict` as Alby suggested with a structure something like:
```
{CartesianTensor: {(device, dtype): ReducedTensorProduct}}
```